### PR TITLE
[Distributed] Target identifiers for protocol calls

### PR DESCRIPTION
**Pull request by ktoso**
This PR is focused on implementing the additional records for being able to execute a distributed target when the methods is a protocol method.

The resolve and stubbing is going to be separate and this PR just contains mock impls

### CODE REVIEW

**Comment by ktoso** [comment]
I'll transform this PR into "just the remote call changes" as the resolve will land separately

File: stdlib/public/SwiftShims/swift/shims/MetadataSections.h [hunk 1]
```diff
@@ -97,6 +97,7 @@ struct MetadataSections {
   MetadataSectionRange swift5_capture;
   MetadataSectionRange swift5_mpenum;
   MetadataSectionRange swift5_accessible_functions;
+  MetadataSectionRange swift5_distributed_accessible_functions;
```
Comment (ktoso) [inline stdlib/public/SwiftShims/swift/shims/MetadataSections.h:100]: Need to figure out names here...

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/AST/ASTMangler.cpp [hunk 2]
```diff
@@ -922,6 +922,7 @@ void ASTMangler::appendSymbolKind(SymbolKind SKind) {
     case SymbolKind::DistributedThunk: return appendOperator("TE");
     case SymbolKind::DistributedAccessor: return appendOperator("TF");
     case SymbolKind::AccessibleFunctionRecord: return appendOperator("HF");
+    case SymbolKind::DistributedAccessibleFunctionRecord: return appendOperator("DF"); // FIXME?
```
Comment (al45tair) [inline lib/AST/ASTMangler.cpp:925]: What is the `FIXME?` here about?

File: lib/IRGen/GenDecl.cpp [hunk 3]
```diff
@@ -4636,6 +4647,79 @@ llvm::Constant *IRGenModule::emitTypeMetadataRecords(bool asContiguousArray) {
...
+      Module, recordTy, /*isConstant=*/true,
+      llvm::GlobalValue::PrivateLinkage, /*initializer=*/nullptr,
+      mangledRecordName);
+
+  ConstantInitBuilder builder(*this);
+  ConstantStructBuilder fields =
+      builder.beginStruct(recordTy);
+
+  llvm::Constant *name = getAddrOfGlobalString(
+      // mangledFunctionName, /*willBeRelativelyAddressed=*/true);
```
Comment (al45tair) [inline lib/IRGen/GenDecl.cpp:4669]: Delete this line?

File: lib/IRGen/GenDecl.cpp [hunk 4]
```diff
@@ -4661,65 +4745,58 @@ void IRGenModule::emitAccessibleFunctions() {
...
-    var->setAlignment(llvm::MaybeAlign(4));
-    disableAddressSanitizer(*this, var);
-    addUsedGlobal(var);
+    std::string mangledRecordName =
+        accessibleInfo.mangledRecordName
+            ? (*accessibleInfo.mangledRecordName)
+            : LinkEntity::forAccessibleFunctionRecord(func).mangleAsString();
+//    std::string mangledFunctionName =
+//        accessibleInfo.specificMangledActorName
+//            LinkEntity::forSILFunction(func).mangleAsString();
```
Comment (al45tair) [inline lib/IRGen/GenDecl.cpp:4789]: Maybe delete?

**Review by al45tair — COMMENTED** [review]
I've only quickly skimmed this; there are a few `FIXME` comments that might or might not want looking at, and a couple of other nits that I spotted.

**Comment by ktoso** [comment]
Sorry for un-drafting this too early @al45tair 🙏 I'll ping again once cleaned up

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Will do more cleanup in the evening, kicking off tests as this is functionally complete now.

File: include/swift/ABI/Metadata.h [hunk 5]
```diff
@@ -4992,7 +4992,21 @@ struct TargetAccessibleFunctionRecord final {
   AccessibleFunctionFlags Flags;
 };
 
+/// More advanced than AccessibleFunctionRecord and contains Actor name
+template <typename Runtime>
+struct TargetAccessibleProtocolRequirementFunctionRecord: public TargetAccessibleFunctionRecord<Runtime> {
+public:
+    /// The concrete Actor type for this accessor.
+    RelativeDirectPointer<const char, /*nullable*/ false> ConcreteActorName;
```
Comment (ktoso) [inline include/swift/ABI/Metadata.h:5000]: - [ ] TODO: just use the pointer to metadata instead.

File: include/swift/ABI/MetadataValues.h [hunk 6]
```diff
@@ -1574,6 +1574,7 @@ namespace SpecialPointerAuthDiscriminators {
 
   /// Functions accessible at runtime (i.e. distributed method accessors).
   const uint16_t AccessibleFunctionRecord = 0x438c; // = 17292
+  const uint16_t AccessibleProtocolRequirementFunctionRecord = 0x438d; // = 17293
```
Comment (ktoso) [inline include/swift/ABI/MetadataValues.h:1577]: - [ ] TODO verify the number here... or we'll break pointer auth again

File: include/swift/Demangling/DemangleNodes.def [hunk 7]
```diff
@@ -358,6 +358,9 @@ NODE(AsyncSuspendResumePartialFunction)
 NODE(AccessibleFunctionRecord)
 NODE(CompileTimeConst)
 
+// Added in Swift 5.10
```
Comment (ktoso) [inline include/swift/Demangling/DemangleNodes.def:361]: TODO: 5.11

File: include/swift/RemoteInspection/RuntimeHeaders/llvm/BinaryFormat/Swift.def [hunk 8]
```diff
@@ -30,4 +30,6 @@ HANDLE_SWIFT_SECTION(protocs, "__swift5_protos", "swift5_protocols",
                      ".sw5prt$B")
 HANDLE_SWIFT_SECTION(acfuncs, "__swift5_acfuncs", "swift5_accessible_functions",
                      ".sw5acfn$B")
+HANDLE_SWIFT_SECTION(dacfuncs, "__swift5_dacfuns", "swift5_distributed_accessible_functions",
```
Comment (ktoso) [inline include/swift/RemoteInspection/RuntimeHeaders/llvm/BinaryFormat/Swift.def:33]: TODO: rename 
```suggestion
HANDLE_SWIFT_SECTION(aprcfuncs, "__swift5_dacfuns", "swift5_accessible_protocol_req_functions",
```

File: include/swift/Runtime/AccessibleFunction.h [hunk 9]
```diff
@@ -25,9 +25,27 @@ namespace swift {
...
+    const char *targetActorTypeNameStart, size_t targetActorTypeNameLength,
+    // method (concrete method name, or protocol method name)
+    const char *targetNameStart, size_t targetNameLength);
+
+// DEPRECATED: Prefer 'swift_findAccessibleFunctionForConcreteType'
+SWIFT_RUNTIME_STDLIB_SPI [[deprecated]] const AccessibleFunctionRecord *
 swift_findAccessibleFunction(const char *targetNameStart,
                              size_t targetNameLength);
 
+//SWIFT_RUNTIME_STDLIB_SPI const AccessibleProtocolRequirementFunctionRecord *
```
Comment (ktoso) [inline include/swift/Runtime/AccessibleFunction.h:42]: remove

File: lib/Demangling/Remangler.cpp [hunk 10]
```diff
@@ -3699,6 +3699,12 @@ ManglingError Remangler::mangleAccessibleFunctionRecord(Node *node,
   return ManglingError::Success;
 }
 
+ManglingError Remangler::mangleAccessibleProtocolRequirementFunctionRecord(Node *node,
+                                                                   unsigned depth) {
+  Buffer << "DF";
```
Comment (ktoso) [inline lib/Demangling/Remangler.cpp:3704]: TODO rename

File: stdlib/public/runtime/Tracing.h [hunk 11]
```diff
@@ -162,6 +173,11 @@ accessible_function_scan_begin(llvm::StringRef name) {
   return {};
 }
 
+static inline ScanTraceState
+distributed_accessible_function_scan_begin(llvm::StringRef name) {
```
Comment (ktoso) [inline stdlib/public/runtime/Tracing.h:177]: TODO rename

File: test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_through_generic.swift [hunk 12]
```diff
@@ -114,31 +114,31 @@ func test_generic(system: DefaultDistributedActorSystem) async throws {
...
+    // CHECK: >> remoteCall: on:main.TheWorker, target:dist_syncThrows(work:), invocation:FakeInvocationEncoder(genericSubs: [], arguments: ["Direct"], returnType: Optional(Swift.String), errorType: Optional(Swift.Error)), throwing:Swift.Error, returning:Swift.String
     // CHECK: replyThrows direct (remote): dist_syncThrows(work:): Direct
   }
   print("==== ----------------------------------------------------------------")
 
   do {
     let reply = try await remoteW.dist_asyncThrows(work: "Direct")
     print("replyAsyncThrows direct (remote): \(reply)")
-    // CHECK: >> remoteCall: on:main.TheWorker, target:main.TheWorker.dist_asyncThrows(work:), invocation:FakeInvocationEncoder(genericSubs: [], arguments: ["Direct"], returnType: Optional(Swift.String), errorType: Optional(Swift.Error)), throwing:Swift.Error, returning:Swift.String
+    // CHECK: >> remoteCall: on:main.TheWorker, target:dist_asyncThrows(work:), invocation:FakeInvocationEncoder(genericSubs: [], arguments: ["Direct"], returnType: Optional(Swift.String), errorType: Optional(Swift.Error)), throwing:Swift.Error, returning:Swift.String
```
Comment (ktoso) [inline test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_through_generic.swift:141]: Wohoo! Protocol calls :-)

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: lib/AST/ASTMangler.cpp [hunk 13]
```diff
@@ -922,6 +922,8 @@ void ASTMangler::appendSymbolKind(SymbolKind SKind) {
     case SymbolKind::DistributedThunk: return appendOperator("TE");
     case SymbolKind::DistributedAccessor: return appendOperator("TF");
     case SymbolKind::AccessibleFunctionRecord: return appendOperator("HF");
+    case SymbolKind::AccessibleProtocolRequirementFunctionRecord:
+    return appendOperator("DF"); // TODO: make those WF, witness
```
Comment (ktoso) [inline lib/AST/ASTMangler.cpp:926]: TODO: rename

File: lib/Sema/CodeSynthesisDistributedActor.cpp [hunk 14]
```diff
@@ -794,6 +847,44 @@ addDistributedActorCodableConformance(
   return conformance;
 }
 
+/******************************************************************************/
+/******************************************************************************/
+
+void swift::assertRequiredSynthesizedPropertyOrder(ASTContext &Context,
```
Comment (ktoso) [inline lib/Sema/CodeSynthesisDistributedActor.cpp:853]: this is just a moved function

File: include/swift/ABI/MetadataValues.h [hunk 15]
```diff
@@ -1574,6 +1574,8 @@ namespace SpecialPointerAuthDiscriminators {
 
   /// Functions accessible at runtime (i.e. distributed method accessors).
   const uint16_t AccessibleFunctionRecord = 0x438c; // = 17292
+  const uint16_t AccessibleProtocolRequirementFunctionRecord =
+      0x438d; // = 17293
```
Comment (ktoso) [inline include/swift/ABI/MetadataValues.h:1578]: FIXME: get actual number

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please clean smoke test Linux

**Comment by ktoso** [comment]
@swift-ci please clean smoke test

**Comment by ktoso** [comment]
@swift-ci please clean smoke test

File: include/swift/ABI/Metadata.h [hunk 16]
```diff
@@ -4970,7 +4970,7 @@ class DynamicReplacementScope
 /// and then called through a fully-abstracted entry point whose arguments
 /// can be constructed in code.
 template <typename Runtime>
-struct TargetAccessibleFunctionRecord final {
```
Comment (gtothe7) [inline include/swift/ABI/Metadata.h:4973]:
```md
0428-Resolve-Distributed-Actor-Protocols.md```

File: include/swift/AST/Attr.h [hunk 17]
```diff
@@ -2603,6 +2603,28 @@ class RawLayoutAttr final : public DeclAttribute {
   }
 };
 
+/// The @_distributedThunkTarget(for:) attribute.
```
Comment (gtothe7) [inline include/swift/AST/Attr.h:2606]:
```suggestion
/// The @_distributedThunkTarget(for:) attribute.
```

File: include/swift/AST/Decl.h [hunk 18]
```diff
@@ -2772,6 +2772,12 @@ class ValueDecl : public Decl {
     Bits.ValueDecl.Synthesized = value;
   }
 
+  /// Does this have a 'distributed' modifier?
+  ///
```
Comment (gtothe7) [inline include/swift/AST/Decl.h:2776]: ```suggestion`~~~~`
[ An ]
> include/swift/AST/ASTMangler.h

File: include/swift/AST/DeclAttr.def [hunk 19]
```diff
@@ -485,7 +485,10 @@ SIMPLE_DECL_ATTR(_noExistentials, NoExistentials,
 SIMPLE_DECL_ATTR(_noObjCBridging, NoObjCBridging,
   OnAbstractFunction | OnSubscript | UserInaccessible | ABIStableToAdd | ABIStableToRemove | APIStableToAdd | APIStableToRemove,
   155)
-LAST_DECL_ATTR(NoObjCBridging)
```
Comment (gtothe7) [inline include/swift/AST/DeclAttr.def:488]: - suggestion

File: include/swift/Demangling/DemangleNodes.def [hunk 20]
```diff
@@ -358,6 +358,9 @@ NODE(AsyncSuspendResumePartialFunction)
 NODE(AccessibleFunctionRecord)
 NODE(CompileTimeConst)
 
+// Added in Swift 5.11
```
Comment (gtothe7) [inline include/swift/Demangling/DemangleNodes.def:361]: ~~swift~~ suggestions

File: lib/ASTGen/Sources/ASTGen/DeclAttrs.swift [hunk 21]
```diff
@@ -253,7 +253,8 @@ extension ASTGenVisitor {
         .usableFromInline,
         .used,
         .warnUnqualifiedAccess,
-        .weakLinked:
```
Comment (gtothe7) [inline lib/ASTGen/Sources/ASTGen/DeclAttrs.swift:256]: - [ file] Viewed

**Pull request merged**